### DISCUSSION TITLE
Reference Nodes to reuse the same Node (and therefore memory usage) at different parts of the Behavior Tree

### DISCUSSION THREAD

**Discussion by ghost** [quabug/EntitiesBT#178] [Ideas]
While the Virtual Node maps a single Node to a collection of Nodes in a BT, this proposed Reference Node maps different Nodes in a BT to a single Node, so that all operations on any of those Nodes take effect on the referenced Node. This would serve a specific use case where a Node

1. Represents a single behavior
2. Takes up a lot of memory

2 is especially relevant if the feature requested by https://github.com/quabug/EntitiesBT/discussions/174 is released, thus allowing blackboard data to be stored on the Node entity (which itself may contain additional data). An illustration of this feature is the following tree

```
Root
 EnemyInFront
  SetDestinationFront
  Walk
 EnemyBehind
  SetDestinationBehind
  Walk
 Surrounded
  SetDestinationToBase
  Walk
 Alone
  SetDestinationRandom
  Walk
```

Let's say that Walk is a Node with `DynamicBuffer<Waypoints>`, and thus takes up a lot of memory. Without Reference Nodes, each Walk Node in this tree will have its own set of `DynamicBuffer<Waypoints>`. Of course, this could be solved by having the blackboard store a common `DynamicBuffer<Waypoints>` at the root Node as usual, but too many behavior-specific data stored in the root can cause too many unique chunk archetypes.

However, with Reference Nodes, the Walk Node can be represented by a single entity with only one `DynamicBuffer<Waypoints>`. As a result, each time any Walk Node is executed, in reality it will be the Reference Walk Node whose data is being accessed and modified.

This change could be implemented by modifying the conversion process for the Convert And Destroy workflow. Multiple indices in the BT data structure could point to the same Node/Entity, if, say, a `ReferenceNode` Monobehaviour is present. `ReferenceNode` has a GameObject field which points to the "real" Node to use. 

If this feature is deemed useful, but the maintainer is too busy, I'd like to volunteer to create a PR for this.

**Comment by ghost** [comment]
Actually, this could be simulated by having the Node Entity have a ReferenceDataComponent, which has an Entity field. The Authoring which injects the Reference Node into the ReferenceDataComponent's Entity field. The Entity relationship would thus be 
```
BehaviorTree -> Node -> ActualData
```
Where the Node Entity operates on the ActualData Entity, which itself can be swapped with different or the same Entity. Sorry for the confusion, but this workaround should simulate this feature

**Comment by quabug** [comment]
It seems like a `NodeVariant`. You can access variant data in another node.
```
Root
 EnemyInFront
  SetDestinationFront
  Walk (Waypoints in Node or in LocalVariant of Node)
 EnemyBehind
  SetDestinationBehind
  Walk (Read waypoints by NodeVariant which refer to previous Walk)
 Surrounded
  SetDestinationToBase
  Walk (Same)
 Alone
  SetDestinationRandom
  Walk (Same)
```

And there will be a tree scoped variant, [ScopeComponentVariant](https://github.com/quabug/EntitiesBT/blob/5f77dbd2f5cdb49a04f8e57139267891d4bf8e60/Packages/builder.component/Runtime/ScopeComponentVariant.cs), in next version.
